File: Contents/code/python/chapter7/either_monad.py
from typing import TypeVar, Generic, Callable

L = TypeVar('L') # Left type (Error)
R = TypeVar('R') # Right type (Success)
R2 = TypeVar('R2')

class Either(Generic[L, R]):
    def __init__(self, is_left: bool, left_val: L = None, right_val: R = None):
        self.is_left = is_left
        self.left_val = left_val
        self.right_val = right_val

    @staticmethod
    def Left(val: L) -> 'Either[L, R]':
        return Either(True, left_val=val)

    @staticmethod
    def Right(val: R) -> 'Either[L, R]':
        return Either(False, right_val=val)

    def bind(self, f: Callable[[R], 'Either[L, R2]']) -> 'Either[L, R2]':
        if self.is_left:
            # Short-circuit, retaining the error
            return Either.Left(self.left_val)
        return f(self.right_val)
        
    def __repr__(self):
        if self.is_left:
            return f"Left({self.left_val})"
        return f"Right({self.right_val})"

# Usage example chaining operations that might fail
def divide_by(y: int) -> Callable[[int], Either[str, int]]:
    def inner(x: int) -> Either[str, int]:
        if y == 0:
            return Either.Left("Cannot divide by zero")
        return Either.Right(x // y)
    return inner

# Chaining successful path: 100 / 2 / 5
success_chain = Either.Right(100).bind(divide_by(2)).bind(divide_by(5))
print(success_chain) # Right(10)

# Chaining failing path: 100 / 0 / 5
fail_chain = Either.Right(100).bind(divide_by(0)).bind(divide_by(5))
print(fail_chain) # Left(Cannot divide by zero)

File: Contents/code/haskel/chapter7/either_monad.hs
-- In Haskell, Either is defined as: data Either a b = Left a | Right b
-- It is natively a Monad where Right is success and Left is failure.

divideBy :: Int -> Int -> Either String Int
divideBy 0 _ = Left "Cannot divide by zero"
divideBy y x = Right (x `div` y)

-- Using the bind operator (>>=)
successChain :: Either String Int
successChain = Right 100 >>= divideBy 2 >>= divideBy 5
-- Result: Right 10

failChain :: Either String Int
failChain = Right 100 >>= divideBy 0 >>= divideBy 5
-- Result: Left "Cannot divide by zero"

-- Using do-notation
doChain :: Either String Int
doChain = do
    x <- Right 100
    y <- divideBy 2 x
    divideBy 5 y

main :: IO ()
main = do
    print successChain
    print failChain
    print doChain
